[RDY TO REVIEW] Ubuntu Debian package

Baaaaam opened the pull request:
This is not ready for merge yet. I PR it to have some discussion and get feedback...

This PR aims to do: Generate Debian package for Cyclus and upload it on dory.

This PR contains:
- 2 new fils/folder in docker:
 - release-14.04/Dockerfile
 - release-16.04/Dockerfile

Those file aims to build Cyclus deb package built respectively on ubuntu 14.04 and ubuntu 16.04 using docker.
 
in the `circle.yml` file, the `ci-testing` part should be integrated to `master` branch part **BEFORE** merging!

I have started [a discussion about this](https://groups.google.com/forum/#!topic/cyclus-dev/USfRJ9C64TE), plz feel free to add any point you think is important!

scopatz commented on CMakeLists.txt:
```diff
@@ -337,11 +337,11 @@ IF(NOT CYCLUS_DOC_ONLY)
-    SET(CPACK_PACKAGE_VERSION_MINOR "0")
-    #SET(CPACK_PACKAGE_VERSION_PATCH "0") # should use commit number here  ?
+    SET(CPACK_PACKAGE_VERSION_MINOR "4")
+    SET(CPACK_PACKAGE_VERSION_PATCH "0") # should use commit number here  ?
```
My answer to the comment here is "no" :)

Baaaaam commented:
I think I finish my work there.

In this PR:
- add 2 docker folder & file for the generation of `cyclus` debian package on ubuntu 14.04 and ubuntu 16.04 
- update the CmakeList.txt, to have the proper generation of the debian package, named as follow: `cyclus-x.x.x_GitHash-ubuntu-??.04.deb`
- update the circle.yml to launch sequentially the debian generation after the update of the cyclus/cuclus:stable docker container, when merging on master.


I still have few suggestions/questions about the organisation of the docker folder/file:
As I PR on cycamore I would like to organize the cocker folder as follow:
  - cyclus/
    - docker/
      - develop-ci
      - release-ci/
        - ubuntu.14.04/
        - ubuntu 16.04/

In the case of yclus we don't need a `master-ci` as everything is happening in CircleCI: it just regenerate the `cyclus/cyclus:lastest` docker container and push it as the `cyclus/cyclus:stable` one.

Baptiste

gonuke commented:
Can you recommend what you think it will take to review this?  (I ask because when we are dealing with detailed build options and configurations, the expertise is generally very localized.)  For example, I would be inclined to attempt to use the DEB file on a clean image to test it.  Is that sufficient? recommended? required?

gonuke commented on CMakeLists.txt:
```diff
@@ -309,13 +309,13 @@ IF(NOT CYCLUS_DOC_ONLY)
         )
 
-    add_custom_target(uninstall
+      ADD_CUSTOM_TARGET(uninstall
```
extra spaces?

gonuke commented on circle.yml:
```diff
@@ -22,5 +22,24 @@ deployment:
+      - docker tag cyclus/cyclus:latest cyclus/cyclus:stable # creation
+      - docker push cyclus/cyclus:stable # push to docker depot
+        ## DEB GENERATION
+          #ubuntu 14.04
```
Does YAML support functions to reuse code?  The 14 and 16 code are identical except for the "14" or "16"

gonuke replied:
Or a loop?  Are these just a list of valid shell commands?

gonuke commented on circle.yml:
```diff
@@ -22,5 +22,24 @@ deployment:
+      - docker push cyclus/cyclus:stable # push to docker depot
+        ## DEB GENERATION
+          #ubuntu 14.04
+      - rm Dockerfile && ln -s docker/release-14.04/Dockerfile # place the correct Dockerfile
```
Since these two dockerfiles are nearly identical....
Instead of linking to different copies, what about using sed to generate each one from a template by substituting in either "14" or "16"?

Baaaaam commented:
Verifying the DEB file are working properly is good, you can in the same time check if the installation of the deb fails when your installation did not match the requirement.

The main problem is when I set this up, I have to set the circle.yml to do it on a specific branch (other than `master` — here `ci-testing`) and when it works I move it to the `master` branch.
But I can’t check if it is working on the master branch, as I need a merged PR on the `master` branch to check it…

Otherwise, feed back about the folder organisation, and the name of the deb…

> Le 29 juil. 2016 à 11:54, Paul Wilson <notifications@github.com> a écrit :
> 
> Can you recommend what you think it will take to review this? (I ask because when we are dealing with detailed build options and configurations, the expertise is generally very localized.) For example, I would be inclined to attempt to use the DEB file on a clean image to test it. Is that sufficient? recommended? required?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/cyclus/cyclus/pull/1259#issuecomment-236233869>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AOcZOtZELpbgBp1FKrXL4jSlMxV195N8ks5qajBfgaJpZM4JMyD1>.
>

Baaaaam commented:
I'll try to implement your suggestion.
 I have to move back this on "Not ready to merge" as I need to set the circle.yml to act on commit on a other branch than the `master` one.

Baaaaam commented:
I write a script which:
- build the proper Dockerfile from the Dockerfile_sample, (using sed)
- load the proper ubuntu docker container
- get cyclus and generate de debian package
- upload it on dory

the script take a ubuntu version as parameter : 12, 14, 16.... and build the deb package on the LTS one (xx.04).

I also modify `circle.yml` accordingly

Baaaaam commented:
I find a bug in the deg generation (sorry about it...)

gonuke commented on docker/release-ci/Dockerfile_sample:
```diff
@@ -0,0 +1,37 @@
+FROM ubuntu:XX.04
```
These template variables should avoid name collisions: perhaps CYCLUS_UBUNTU_MAJOR_VERSION, CYCLUS_REPO_HASH

gonuke commented on docker/release-ci/Dockerfile_sample:
```diff
@@ -0,0 +1,37 @@
+WORKDIR cyclus
+RUN mkdir -p deb
+WORKDIR /cyclus/deb
+RUN cmake -DCMAKE_INSTALL_PREFIX=/usr/local -DCPACK_SYSTEM_NAME=HASH-ubuntu.XX.04 ..
```
Separate the HASH from the distribution name and assemble them in the CMake file (separation of concerns).

Baaaaam commented:
I have applied the different comments:
- build a script taking ubuntu version as parameter, which build cyclus against, it make the deb and upload it on dory, (the script also get the hash commit number and put in on the deb package name)
- put explicit template name on variables to avoid name collisions
- separate HASH number and system_name from the building comment...

gonuke commented:
One small last request: rename `Dockerfile_sample` to `Dockert_template`.  This is not a sample file but a template to be finalized through the sed commands.

Baaaaam commented:
just figure out that deb did not have dependencies requirement...

Baaaaam commented on docker/release-ci/Dockerfile_template:
```diff
@@ -0,0 +1,37 @@
+    python \
+    python-tables \
+    python-numpy \
+    python-nose
```
Can't use cyclus-deps, because we need to control the ubuntu version.
but this should be documented...

Baaaaam commented on docker/deb-ci/Dockerfile_template:
```diff
@@ -0,0 +1,37 @@
+WORKDIR cyclus
+RUN mkdir -p deb
+WORKDIR /cyclus/deb
+RUN cmake -DCMAKE_INSTALL_PREFIX=/usr -DCPACK_HASH_NAME=CYCLUS_REPO_HASH -DCPACK_SYSTEM_NAME=ubuntu.CYCLUS_UBUNTU_MAJOR_VERSION.04 ..
```
Despite previously defining  -DCMAKE_INSTALL_PREFIX=/usr/local, cyclus file are installed (when using dpkg -i) in /usr directly...
So cyclus is looking for /usr/local... but don't find its stuff...

when  switch to -DCMAKE_INSTALL_PREFIX=/usr it is working.
So we can either using /usr as install prefix or try to understand why dpkg install cyclus in /usr despite the cpack/cmake specification...

gonuke replied:
You've spent more time on this than me, but maybe this helps: http://stackoverflow.com/questions/6712000/cmake-cpack-package-installation-path-nightmare

Baaaaam commented:
I have reduce the dependencies of cyclus binary to:
libtcmalloc-minimal4 (= 2.1-2ubuntu1.1)

libboost-filesystem1.54.0
libboost-program-options1.54.0
 libhdf5-7
libxml++2.6-2
libboost-serialization1.54.0
coinor-libcbc3

to have the unit test working and been able to launch `cyclus` (without input file) and get a normal error message....
I start to add it in the CMakefile...
BaM

gonuke commented on docker/deb-ci/build_upload_deb.sh:
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
+if [ $# -ne 1 ]; then 
+  echo "illegal number of parameters"
```
This should exit with a failure status as well

Baaaaam commented:
@gonuke this is ready for review:
I have addressed all the issues:
- add a failure exit when the wrong number of parameters is given to the `build_upload_deb.sh`script
- now the dependencies are set and require to install cyclus on ubuntu 14.04 and 16.04 (working on both)
- the installation is done in /usr/local

Unit test are working, and running Cyclus (without input file give correct message)
This require a final test with Cycamore to be sure everything is working as expected

Baaaaam commented:
This should be fine, and ready to merge.

Baaaaam commented:
I have check the deb is working properly on a fresh Ubuntu 14 and ubuntu 16, all installation and `cyclus_unit_tests` are working

gonuke commented on CMakeLists.txt:
```diff
@@ -326,29 +326,50 @@ IF(NOT CYCLUS_DOC_ONLY)
-    #SET(CPACK_PACKAGE_VERSION_PATCH "0") # should use commit number here  ?
-    SET(CPACK_PACKAGE_NAME "${CMAKE_PROJECT_NAME}")
+    SET(CPACK_PACKAGE_VERSION_MINOR "4")
+    SET(CPACK_PACKAGE_VERSION_PATCH "0") 
```
Is there somewhere to add a modifier, like "RC" for a release candidate, to this version string?  (I'm being a little pedantic here)

Baaaaam replied:
I see the point, it is possible.
but how to differentiate automatically a release from a release candidate.

gonuke commented on CMakeLists.txt:
```diff
@@ -326,29 +326,50 @@ IF(NOT CYCLUS_DOC_ONLY)
+    
+    SET(
+      CPACK_PACKAGE_FILE_NAME
+      "${CPACK_PACKAGE_NAME}_${CPACK_PACKAGE_VERSION_MAJOR}.${CPACK_PACKAGE_VERSION_MINOR}.${CPACK_PACKAGE_VERSION_PATCH}_${CPACK_HASH_NAME}_${CPACK_SYSTEM_NAME}"
```
Could this be reworked as 2 or 3 set commands so that each fits on a line?  Is it worth it?

gonuke replied:
Or maybe there is just a way to break lines within the string in a CMAKE file?

gonuke replied:
http://stackoverflow.com/questions/7637539/cmake-how-to-split-strings-across-multiple-lines

gonuke commented on docker/deb-ci/Dockerfile_template:
```diff
@@ -0,0 +1,37 @@
+WORKDIR cyclus
+RUN mkdir -p deb
+WORKDIR /cyclus/deb
+RUN cmake -DCPACK_PACKAGING_INSTALL_PREFIX=/usr/local -DCMAKE_INSTALL_PREFIX=/usr/local -DCPACK_HASH_NAME=CYCLUS_REPO_HASH -DCPACK_SYSTEM_NAME=ubuntu.CYCLUS_UBUNTU_MAJOR_VERSION.04 ..
```
Is there a reason to prefer `/usr/local` over other locations?  e.g. `/opt`?  Is this the "Ubuntu-ish" way to do things?

Baaaaam replied:
There is a discussion about it: 
http://unix.stackexchange.com/questions/30/where-should-i-put-software-i-compile-myself

I think people agree that `/usr` is for the distribution stuff, `/usr/local` for the safe, and `/opt` for you own compiled tool.
According to that, we should `cyclus` in `/opt`...
But /opt is not in the $PATH by default on `ubuntu`, so I think I prefer to keep the installation in `/usr/local` to simplify users installation...

Baaaaam commented:
I move the trigger to ci-testing again to check the last modification, plz do not merge yet

Baaaaam commented:
I have recheck deb generation with all the PR, it seems to be working.
the deb package van be found [there](http://dory.fuelcycle.org:4848)

Unless you have comment, this could be merged as is (trigger of CI is on `master` branch).

gonuke commented:
On 16.04, cyclus reports a very long dependency list,  much longer than on 14.04.... this isn't a show stopper, but I wonder if we know why?

16.04:
```
root@server-02:~# dpkg -I cyclus_1.4.0_2d902fd_ubuntu.16.04.deb 
 new debian package, version 2.0.
 size 28351448 bytes: control archive=5885 bytes.
     833 bytes,    10 lines      control              
   14014 bytes,   182 lines      md5sums              
 Package: cyclus
 Version: 1.4.0
 Section: devel
 Priority: optional
 Architecture: amd64
 Depends: coinor-libcbc3, coinor-libclp1, coinor-libcoinutils3v5, coinor-libosi1v5, libboost-filesystem1.58.0, libboost-program-options1.58.0, libboost-serialization1.58.0, libboost-system1.58.0, libc6 (>= 2.14), libgcc1 (>= 1:3.0), libglibmm-2.4-1v5 (>= 2.46.0), libhdf5-10, libsqlite3-0 (>= 3.5.9), libstdc++6 (>= 5.2), libtcmalloc-minimal4, libxml++2.6-2v5 (>= 2.40.0), libxml2 (>= 2.7.4), libtcmalloc-minimal4 (>= 2.1), libboost-filesystem-dev (>= 1.54.0), libboost-program-options-dev (>= 1.54.0), libboost-serialization-dev (>= 1.54.0), libhdf5-dev (>= 1.8.11), libxml++2.6-dev (>= 2.36.0), coinor-libcbc-dev (>= 2.8.7)
 Installed-Size: 72754
 Maintainer: CNERG, UW-Madison
 Description: Cyclus : A nuclear fuel cycle simulator from UW-Madison.
```

14.04:
```
root@server-02:~# dpkg -I cyclus_1.4.0_2d902fd_ubuntu.14.04.deb 
 new debian package, version 2.0.
 size 27911328 bytes: control archive=5758 bytes.
     450 bytes,    10 lines      control              
   14014 bytes,   182 lines      md5sums              
 Package: cyclus
 Version: 1.4.0
 Section: devel
 Priority: optional
 Architecture: amd64
 Depends: libtcmalloc-minimal4 (>= 2.1), libboost-filesystem-dev (>= 1.54.0), libboost-program-options-dev (>= 1.54.0), libboost-serialization-dev (>= 1.54.0), libhdf5-dev (>= 1.8.11), libxml++2.6-dev (>= 2.36.0), coinor-libcbc-dev (>= 2.8.7)
 Installed-Size: 69662
 Maintainer: CNERG, UW-Madison
 Description: Cyclus : A nuclear fuel cycle simulator from UW-Madison.
```

gonuke commented:
I'm ready to merge this if:
* we can make 16.04 have a smaller dependency list, OR
* we decide that it is not possible to do that...

Baaaaam commented:
I am not sure about it, I am trying to investigate it

Baaaaam commented on CMakeLists.txt:
```diff
@@ -383,7 +383,7 @@ IF(NOT CYCLUS_DOC_ONLY)
     SET(CPACK_INCLUDE_TOPLEVEL_DIRECTORY "1")
-    SET(CPACK_DEBIAN_PACKAGE_SHLIBDEPS ON)
+    SET(CPACK_DEBIAN_PACKAGE_SHLIBDEPS OFF)
     CONFIGURE_FILE(
```
this reduce the large dependency list on ubuntu 16.04.

gonuke commented:
Thanks a tonne @Baaaaam!!!  Merging this now
